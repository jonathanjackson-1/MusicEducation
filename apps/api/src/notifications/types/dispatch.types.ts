import { NotificationChannel } from './notification-channel.enum';

export interface ChannelDispatchRequest {
  recipient: string;
  payload: Record<string, unknown>;
}

export interface ChannelDispatchReceipt {
  channel: NotificationChannel;
  provider: string;
  recipient: string;
  status: 'queued' | 'sent' | 'skipped' | 'failed';
  metadata?: Record<string, unknown>;
}
