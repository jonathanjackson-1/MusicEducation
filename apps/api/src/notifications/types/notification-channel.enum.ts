export enum NotificationChannel {
  EMAIL = 'EMAIL',
  SMS = 'SMS',
  PUSH = 'PUSH',
}

export const NOTIFICATION_CHANNELS = Object.values(NotificationChannel);
