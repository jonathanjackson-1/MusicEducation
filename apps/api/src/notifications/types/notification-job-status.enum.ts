export enum NotificationJobStatus {
  PENDING = 'PENDING',
  QUEUED = 'QUEUED',
  SENT = 'SENT',
  FAILED = 'FAILED',
  CANCELLED = 'CANCELLED',
}
