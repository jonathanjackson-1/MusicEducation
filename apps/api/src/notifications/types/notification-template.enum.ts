export enum NotificationTemplateType {
  BOOKING_CONFIRMATION = 'BOOKING_CONFIRMATION',
  BOOKING_RESCHEDULE_PROPOSAL = 'BOOKING_RESCHEDULE_PROPOSAL',
  BOOKING_CANCELLATION = 'BOOKING_CANCELLATION',
  ASSIGNMENT_CREATED = 'ASSIGNMENT_CREATED',
  SUBMISSION_FEEDBACK = 'SUBMISSION_FEEDBACK',
  PRACTICE_STREAK = 'PRACTICE_STREAK',
}

export const NOTIFICATION_TEMPLATES = Object.values(NotificationTemplateType);
