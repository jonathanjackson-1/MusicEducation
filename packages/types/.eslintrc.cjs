module.exports = {
  root: true,
  extends: ["@soundstudio/config/eslint/base"],
  parserOptions: {
    tsconfigRootDir: __dirname,
    project: ["./tsconfig.json"]
  }
};
