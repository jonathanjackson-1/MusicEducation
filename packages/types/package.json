{
  "name": "@soundstudio/types",
  "version": "0.1.0",
  "private": false,
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    }
  },
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "build": "tsc -p tsconfig.build.json",
    "lint": "eslint 'src/**/*.{ts,tsx}'",
    "typecheck": "tsc -p tsconfig.json --noEmit",
    "test": "vitest"
  },
  "devDependencies": {
    "@soundstudio/config": "workspace:*",
    "@typescript-eslint/eslint-plugin": "^6.19.1",
    "@typescript-eslint/parser": "^6.19.1",
    "eslint": "^8.56.0",
    "eslint-config-prettier": "^9.1.0",
    "eslint-import-resolver-typescript": "^3.6.1",
    "eslint-plugin-import": "^2.29.1",
    "typescript": "^5.3.3",
    "vitest": "^1.2.2"
  }
}
