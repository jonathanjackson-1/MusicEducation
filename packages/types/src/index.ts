export type LessonLevel = 'beginner' | 'intermediate' | 'advanced';

export type UserRole = 'student' | 'teacher' | 'admin';

export interface User {
  id: string;
  email: string;
  name: string;
  role: UserRole;
  avatarUrl?: string | null;
  joinedAt: string;
  timezone?: string | null;
  preferredInstrument?: string | null;
}

export interface LessonSummary {
  id: string;
  title: string;
  level: LessonLevel;
  durationMinutes: number;
  thumbnailUrl?: string | null;
}

export interface LessonResource {
  id: string;
  lessonId: string;
  type: 'video' | 'audio' | 'sheet-music' | 'pdf' | 'link';
  title: string;
  url: string;
  durationSeconds?: number;
  description?: string;
}

export interface Lesson extends LessonSummary {
  description: string;
  instructorId: string;
  tags?: string[];
  publishedAt?: string;
  updatedAt?: string;
  resources?: LessonResource[];
  objectives?: string[];
}

export type AssignmentStatus = 'pending' | 'submitted' | 'completed';

export interface Assignment {
  id: string;
  lessonId: string;
  title: string;
  instructions: string;
  dueDate?: string;
  availableAt: string;
  submittedAt?: string;
  status?: AssignmentStatus;
  attachments?: LessonResource[];
}

export type PracticeMood = 'confident' | 'neutral' | 'challenged';

export interface PracticeLog {
  id: string;
  userId: string;
  lessonId?: string;
  assignmentId?: string;
  startedAt: string;
  durationMinutes: number;
  notes?: string;
  mood?: PracticeMood;
  tempoBpm?: number;
  focusAreas?: string[];
}

export interface PracticeStreak {
  userId: string;
  current: number;
  longest: number;
  updatedAt: string;
  startedAt?: string;
}

export type PolicyType = 'terms' | 'privacy' | 'community' | 'safety';

export interface Policy {
  id: string;
  type: PolicyType;
  title: string;
  body: string;
  version: string;
  publishedAt: string;
  updatedAt?: string;
}

export type BillingInterval = 'monthly' | 'quarterly' | 'yearly' | 'lifetime';

export interface Plan {
  id: string;
  name: string;
  description?: string;
  priceCents: number;
  currency: string;
  interval: BillingInterval;
  features: string[];
  mostPopular?: boolean;
  trialDays?: number;
}

export interface ApiPaginationMeta {
  page: number;
  perPage: number;
  totalItems: number;
  totalPages: number;
}

export interface ApiMeta {
  requestId?: string;
  servedAt: string;
  locale?: string;
  pagination?: ApiPaginationMeta;
}

export interface ApiErrorDetail {
  code: string;
  message: string;
  field?: string;
  helpUrl?: string;
}

export interface ApiResponse<TData> {
  data: TData;
  meta?: ApiMeta;
  errors?: ApiErrorDetail[];
}

export interface PaginatedResponse<TItem> extends ApiResponse<TItem[]> {
  meta: ApiMeta & { pagination: ApiPaginationMeta };
}

export type UserResponse = ApiResponse<User>;
export type UsersResponse = PaginatedResponse<User>;

export type LessonResponse = ApiResponse<Lesson>;
export type LessonsResponse = PaginatedResponse<Lesson>;
export type LessonSummariesResponse = PaginatedResponse<LessonSummary>;

export type AssignmentResponse = ApiResponse<Assignment>;
export type AssignmentsResponse = PaginatedResponse<Assignment>;

export type PracticeLogResponse = ApiResponse<PracticeLog>;
export type PracticeLogsResponse = PaginatedResponse<PracticeLog>;

export type PolicyResponse = ApiResponse<Policy>;
export type PoliciesResponse = PaginatedResponse<Policy>;

export type PlanResponse = ApiResponse<Plan>;
export type PlansResponse = ApiResponse<Plan[]>;
