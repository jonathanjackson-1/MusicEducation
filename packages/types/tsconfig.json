{
  "extends": "@soundstudio/config/tsconfig/base",
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "dist",
    "declaration": true,
    "declarationMap": true
  },
  "include": ["src/**/*.ts"],
  "exclude": ["node_modules", "dist"]
}
