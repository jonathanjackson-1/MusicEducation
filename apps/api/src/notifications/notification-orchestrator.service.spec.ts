import { NotificationOrchestratorService } from './notification-orchestrator.service';
import { NotificationTemplateEngine } from './template-engine/notification-template-engine.service';
import { NotificationTemplateType } from './types/notification-template.enum';
import { NotificationChannel } from './types/notification-channel.enum';
import { ChannelDispatchReceipt } from './types/dispatch.types';

class FakeEmailChannel {
  public readonly messages: any[] = [];

  async send(options: any): Promise<ChannelDispatchReceipt> {
    this.messages.push(options);
    return {
      channel: NotificationChannel.EMAIL,
      provider: 'fake-email',
      recipient: options.to,
      status: 'sent',
      metadata: { subject: options.template.subject },
    };
  }
}

class FakeSmsChannel {
  public readonly messages: any[] = [];

  async send(options: any): Promise<ChannelDispatchReceipt> {
    this.messages.push(options);
    return {
      channel: NotificationChannel.SMS,
      provider: 'fake-sms',
      recipient: options.to,
      status: 'sent',
      metadata: {},
    };
  }
}

class FakePushChannel {
  public readonly messages: any[] = [];

  async send(options: any): Promise<ChannelDispatchReceipt[]> {
    this.messages.push(options);
    return options.tokens.map((token: any) => ({
      channel: NotificationChannel.PUSH,
      provider: `fake-${token.provider ?? 'fcm'}`,
      recipient: token.token,
      status: 'sent',
      metadata: token.metadata ?? {},
    }));
  }
}

describe('NotificationOrchestratorService', () => {
  const prisma: any = {
    studio: {
      findUnique: jest.fn().mockResolvedValue({
        id: 'studio-1',
        name: 'Sound Studio',
        locale: 'en',
      }),
    },
    user: {
      findUnique: jest.fn().mockResolvedValue({
        id: 'user-1',
        firstName: 'Avery',
        lastName: 'Student',
        email: 'avery@student.test',
      }),
    },
    notification: {
      create: jest.fn().mockResolvedValue(null),
    },
    notificationPreference: {
      findMany: jest.fn().mockResolvedValue([
        { channel: NotificationChannel.EMAIL, enabled: true },
        { channel: NotificationChannel.SMS, enabled: true },
        { channel: NotificationChannel.PUSH, enabled: true },
      ]),
      upsert: jest.fn(),
    },
    notificationSubscription: {
      findMany: jest.fn().mockResolvedValue([
        {
          channel: NotificationChannel.SMS,
          endpoint: '+15551234567',
          locale: 'en',
        },
        {
          channel: NotificationChannel.PUSH,
          endpoint: 'fcm-token',
          provider: 'fcm',
          metadata: { device: 'android' },
        },
        {
          channel: NotificationChannel.PUSH,
          endpoint: 'apns-token',
          provider: 'apns',
          metadata: { device: 'ios' },
        },
      ]),
      upsert: jest.fn(),
    },
  };

  const emailChannel = new FakeEmailChannel();
  const smsChannel = new FakeSmsChannel();
  const pushChannel = new FakePushChannel();
  const orchestrator = new NotificationOrchestratorService(
    prisma,
    new NotificationTemplateEngine(),
    emailChannel as any,
    smsChannel as any,
    pushChannel as any,
  );

  it('dispatches notifications across channels respecting preferences', async () => {
    const receipts = await orchestrator.dispatch({
      studioId: 'studio-1',
      userId: 'user-1',
      template: NotificationTemplateType.BOOKING_CONFIRMATION,
      context: {
        reminder: true,
        reminderWindow: '24 hours',
        lessonDate: 'April 22, 2024',
        lessonTime: '3:30 PM',
        educatorName: 'Logan Teacher',
        bookingUrl: 'https://studio.example.com/lessons/lesson-123',
        roomName: 'Studio A',
      },
    });

    expect({
      receipts,
      emailMessages: emailChannel.messages,
      smsMessages: smsChannel.messages,
      pushMessages: pushChannel.messages,
      loggedNotification: prisma.notification.create.mock.calls[0][0],
    }).toMatchSnapshot();
  });

  it('subscribes to channels and upserts preferences', async () => {
    prisma.notificationSubscription.upsert.mockResolvedValue({
      id: 'sub-1',
      channel: NotificationChannel.PUSH,
      endpoint: 'device-token',
      provider: 'fcm',
    });
    prisma.notificationPreference.upsert.mockResolvedValue({
      id: 'pref-1',
      channel: NotificationChannel.PUSH,
      template: NotificationTemplateType.PRACTICE_STREAK,
      enabled: true,
    });

    const result = await orchestrator.subscribe({
      studioId: 'studio-1',
      userId: 'user-1',
      channel: NotificationChannel.PUSH,
      endpoint: 'device-token',
      provider: 'fcm',
      locale: 'en',
      metadata: { platform: 'ios' },
      templates: [
        {
          template: NotificationTemplateType.PRACTICE_STREAK,
          enabled: true,
        },
      ],
    });

    expect({
      subscriptionCall: prisma.notificationSubscription.upsert.mock.calls[0][0],
      preferenceCall: prisma.notificationPreference.upsert.mock.calls[0][0],
      result,
    }).toMatchSnapshot();
  });
});
