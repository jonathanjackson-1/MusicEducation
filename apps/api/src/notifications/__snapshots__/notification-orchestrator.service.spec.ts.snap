// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationOrchestratorService dispatches notifications across channels respecting preferences 1`] = `
{
  "emailMessages": [
    {
      "metadata": {
        "locale": "en",
        "template": "BOOKING_CONFIRMATION",
      },
      "template": {
        "html": "<!doctype html><html><body>
  
  <div class="mj-body" style="background-color:#f9fafb;">

    <section style="background-color:#ffffff;padding:24px;">

      <div class="mj-column">

        <p style="font-size:20px;font-weight:600;">
Hello Avery Student,</p>
        <p style="padding-top:12px;">

            This is a friendly reminder that your lesson with Logan Teacher begins in 24 hours on April 22, 2024 at 3:30 PM.
        </p>
          <p style="padding-top:8px;">
Location: Studio A</p>
        <a href="https://studio.example.com/lessons/lesson-123" style="display:inline-block;padding:12px 20px;border-radius:6px;text-decoration:none;padding-top:20px;">View lesson details</a>
        <p style="font-size:14px;color:#6b7280;padding-top:16px;">

          Sound Studio • April 22, 2024 • 3:30 PM
        </p>
      </div>
    </section>
  </div>
</body></html>
",
        "subject": "Reminder: Lesson with Logan Teacher",
        "text": "<!doctype html> Hello Avery Student, This is a friendly reminder that your lesson with Logan Teacher begins in 24 hours on April 22, 2024 at 3:30 PM. Location: Studio A View lesson details Sound Studio • April 22, 2024 • 3:30 PM",
      },
      "to": "avery@student.test",
    },
  ],
  "loggedNotification": {
    "data": {
      "payload": {
        "channels": [
          "EMAIL",
          "SMS",
          "PUSH",
        ],
        "context": {
          "bookingUrl": "https://studio.example.com/lessons/lesson-123",
          "educatorName": "Logan Teacher",
          "lessonDate": "April 22, 2024",
          "lessonTime": "3:30 PM",
          "pushData": {
            "studioId": "studio-1",
            "template": "BOOKING_CONFIRMATION",
            "userId": "user-1",
          },
          "recipientName": "Avery Student",
          "reminder": true,
          "reminderWindow": "24 hours",
          "roomName": "Studio A",
          "studioName": "Sound Studio",
        },
        "locale": "en",
        "receipts": [
          {
            "channel": "EMAIL",
            "metadata": {
              "subject": "Reminder: Lesson with Logan Teacher",
            },
            "provider": "fake-email",
            "recipient": "avery@student.test",
            "status": "sent",
          },
          {
            "channel": "SMS",
            "metadata": {},
            "provider": "fake-sms",
            "recipient": "+15551234567",
            "status": "sent",
          },
          {
            "channel": "PUSH",
            "metadata": {
              "device": "android",
            },
            "provider": "fake-fcm",
            "recipient": "fcm-token",
            "status": "sent",
          },
          {
            "channel": "PUSH",
            "metadata": {
              "device": "ios",
            },
            "provider": "fake-apns",
            "recipient": "apns-token",
            "status": "sent",
          },
        ],
      },
      "template": "BOOKING_CONFIRMATION",
      "type": "BOOKING_CONFIRMATION",
      "userId": "user-1",
    },
  },
  "pushMessages": [
    {
      "payload": {
        "data": {
          "studioId": "studio-1",
          "template": "BOOKING_CONFIRMATION",
          "userId": "user-1",
        },
        "notification": {
          "body": "Lesson with Logan Teacher on April 22, 2024 at 3:30 PM.",
          "title": "Lesson reminder",
        },
      },
      "tokens": [
        {
          "metadata": {
            "device": "android",
          },
          "provider": "fcm",
          "token": "fcm-token",
        },
        {
          "metadata": {
            "device": "ios",
          },
          "provider": "apns",
          "token": "apns-token",
        },
      ],
    },
  ],
  "receipts": [
    {
      "channel": "EMAIL",
      "metadata": {
        "subject": "Reminder: Lesson with Logan Teacher",
      },
      "provider": "fake-email",
      "recipient": "avery@student.test",
      "status": "sent",
    },
    {
      "channel": "SMS",
      "metadata": {},
      "provider": "fake-sms",
      "recipient": "+15551234567",
      "status": "sent",
    },
    {
      "channel": "PUSH",
      "metadata": {
        "device": "android",
      },
      "provider": "fake-fcm",
      "recipient": "fcm-token",
      "status": "sent",
    },
    {
      "channel": "PUSH",
      "metadata": {
        "device": "ios",
      },
      "provider": "fake-apns",
      "recipient": "apns-token",
      "status": "sent",
    },
  ],
  "smsMessages": [
    {
      "body": "Reminder: Lesson with Logan Teacher on April 22, 2024 at 3:30 PM. Room: Studio A. Details: https://studio.example.com/lessons/lesson-123",
      "metadata": {
        "locale": "en",
        "template": "BOOKING_CONFIRMATION",
      },
      "to": "+15551234567",
    },
  ],
}
`;

exports[`NotificationOrchestratorService subscribes to channels and upserts preferences 1`] = `
{
  "preferenceCall": {
    "create": {
      "channel": "PUSH",
      "enabled": true,
      "locale": "en",
      "studioId": "studio-1",
      "template": "PRACTICE_STREAK",
      "userId": "user-1",
    },
    "update": {
      "enabled": true,
      "locale": "en",
    },
    "where": {
      "studioId_userId_template_channel": {
        "channel": "PUSH",
        "studioId": "studio-1",
        "template": "PRACTICE_STREAK",
        "userId": "user-1",
      },
    },
  },
  "result": {
    "preferences": [
      {
        "channel": "PUSH",
        "enabled": true,
        "id": "pref-1",
        "template": "PRACTICE_STREAK",
      },
    ],
    "subscription": {
      "channel": "PUSH",
      "endpoint": "device-token",
      "id": "sub-1",
      "provider": "fcm",
    },
  },
  "subscriptionCall": {
    "create": {
      "active": true,
      "channel": "PUSH",
      "endpoint": "device-token",
      "locale": "en",
      "metadata": {
        "platform": "ios",
      },
      "provider": "fcm",
      "studioId": "studio-1",
      "userId": "user-1",
    },
    "update": {
      "active": true,
      "locale": "en",
      "metadata": {
        "platform": "ios",
      },
      "provider": "fcm",
    },
    "where": {
      "studioId_userId_channel_endpoint": {
        "channel": "PUSH",
        "endpoint": "device-token",
        "studioId": "studio-1",
        "userId": "user-1",
      },
    },
  },
}
`;
